File: lib_python/environment.py
from color import Color
from light_value import LightValue

from .vec import Vec3


class Environment:

    def __init__(
        self,
        light_diffusion_strength: float = 0.99,
        light_algorithm: str = "none",
    ) -> None:

        self.light_diffusion_strength: float = light_diffusion_strength
        self.light_algorithm: str = light_algorithm


class EnvironmentColor(Environment):

    def __init__(
        self,
        light_diffusion_strength: float = 0.99,
        light_algorithm: str = "none",
        environment_color: Color = Color(255, 255, 255),
        environment_color_light_emission: LightValue = LightValue(),
    ) -> None:

        super().__init__(
            light_diffusion_strength,
            light_algorithm
        )

        self.environment_color: Color = environment_color
        self.environment_color_light_emission: LightValue = environment_color_light_emission


class EnvironmentSkyBox(Environment):

    def __init__(
        self,
        light_diffusion_strength: float = 0.99,
        light_algorithm: str = "none",
        sky_color: Color = Color(145, 200, 228),
        sky_color_light_emission: LightValue = LightValue(),
        ground_color: Color = Color(32, 94, 97),
        ground_color_light_emission: LightValue = LightValue(),
        sun_direction: Vec3 = Vec3(0, 0, 0),
        sun_light_emission: LightValue = LightValue(10, 10, 10)
    ) -> None:

        super().__init__(
            light_diffusion_strength,
            light_algorithm
        )

        self.sky_color: Color = sky_color
        self.sky_color_light_emission: LightValue = sky_color_light_emission
        self.ground_color: Color = ground_color
        self.ground_color_light_emission: LightValue = ground_color_light_emission
        self.sun_direction: Vec3 = sun_direction
        self.sun_light_emission: LightValue = sun_light_emission

File: lib_js_naive/environment.js
/**
 * Environment - Background/skybox configuration
 * Matches Python lib_python/environment.py
 */
class Environment {
    constructor(options = {}) {
        this.type = options.type || 'color';
        this.lightDiffusionStrength = options.lightDiffusionStrength || 1.0;
        this.lightAlgorithm = options.lightAlgorithm || 'none';
    }
}

class EnvironmentColor extends Environment {
    constructor(options = {}) {
        super(options);
        this.type = 'color';
        this.color = options.color || new Color(0, 0, 0, 0);
        this.lightEmission = options.lightEmission || 1.0;
    }
}

class EnvironmentSkyBox extends Environment {
    constructor(options = {}) {
        super(options);
        this.type = 'skybox';
        this.skyColor = options.skyColor || new Color(135, 206, 235);
        this.skyColorLightEmission = options.skyColorLightEmission || 0.1;
        this.groundColor = options.groundColor || new Color(30, 77, 43);
        this.groundColorLightEmission = options.groundColorLightEmission || 0.0;
        this.sunDirection = options.sunDirection || new Vec3(0, -1, 0);
        this.sunLightEmission = options.sunLightEmission || 1.0;
    }
}

/**
 * Parse environment from JSON
 * @param {object} json - JSON object
 * @returns {Environment}
 */
function parseEnvironment(json) {
    const type = json.environment_type || 'color';
    const lightDiffusionStrength = json.light_diffusion_strength || 1.0;
    const lightAlgorithm = json.light_algorithm || 'none';

    if (type === 'skybox') {
        return new EnvironmentSkyBox({
            lightDiffusionStrength,
            lightAlgorithm,
            skyColor: parseColor(json.sky_color || '#87CEEB'),
            skyColorLightEmission: json.sky_color_light_emission || 0.1,
            groundColor: parseColor(json.ground_color || '#1E4D2B'),
            groundColorLightEmission: json.ground_color_light_emission || 0.0,
            sunDirection: parseVec3(json.sun_direction || [0, -1, 0]),
            sunLightEmission: json.sun_light_emission || 1.0,
        });
    }

    return new EnvironmentColor({
        lightDiffusionStrength,
        lightAlgorithm,
        color: parseColor(json.environment_color || [0, 0, 0, 0]),
        lightEmission: json.environment_color_light_emission || 1.0,
    });
}

// Expose to window
window.Environment = Environment;
window.EnvironmentColor = EnvironmentColor;
window.EnvironmentSkyBox = EnvironmentSkyBox;
window.parseEnvironment = parseEnvironment;
